### DISCUSSION TITLE
Question about file reader and multiple threads

### DISCUSSION THREAD

**Discussion by rutledj** [nextgenhealthcare/connect#5978] [Q&A]
If I set the Max Threads on a source file reader, is there any way to actually see how many threads are running for that channel?  
Also, for a file reader, is there any way to tell what file it is currently reading? I added a filename and start time in the source transformer, but that doesn't really  give you the time it started reading the file, I was going to add some javascript to the attachment handler to record the start but that is already set to Entire Message to create an attachment.

**Comment by jonbartels** [comment]
> any way to actually see how many threads are running for that channel

Take a thread dump from the jvm. I prefer `jstack` https://www.baeldung.com/java-thread-dump

> Also, for a file reader, is there any way to tell what file it is currently reading?

No. [The code doesn't log its progress on a file, by file basis](https://github.com/nextgenhealthcare/connect/blob/c1bbf981b396279984124adbd50ddb4851b2bc76/server/src/com/mirth/connect/connectors/file/FileReceiver.java#L180). 

File readers (and I think all other polling connectors) run a single thread for the poll, then that single thread can generate multiple messages which can then be processed by the source  connector threads.

**Reply by jonbartels** [reply]
@rutledj what problem are you trying to solve?

Is your file reader reading a local filesystem, SFTP, S3, or a netmount?

**Comment by rutledj** [comment]
Trying to see what the speed is for a file reader to read in files from an smb drive, destination is a file writer to a mapped network drive.
 For some reason, it appears to be very slow. 

So, it sounds like increasing the source threads only affects it processing through the channel. The channel is set to move the message as an attachment. No processing happening with the message. Just moving them from one place to another.

So doesn't sound like increasing threads will do anything.

**Reply by jonbartels** [reply]
With an SMB mount on one end and a mapped drive on the other, latency over the network to the filesystem would be the first and most likely performance hit.

The common playbook is:
1. Use locally mounted drives (netmount or mapped drives) - This doesn't escape the latency but shifts it from Mirth to your OS which is probably better able to manage it 
2. Tune your file and directory patterns. There is a known, unresolved bug with SFTP and how it uses `ls`. There *may* be similar issues with SMB. Tuning your patterns to be very explicit will help.
3. If your workflow tolerates it, turn on destination queues. Let your source reader read then let your destination queue and send. This will help if your destination is slower than your reader. *Be warned that if your destination isn't done with the batch and queues are on, that the source can re-read files under certain circumstances. Read and understand what queues do to move-to directories and so on.*
4. If none of the above helps, change your reading patterns. Is there a DB table or other computed value that will let you read and manage the file statuses and locations more precisely so you read one file instead of doing pattern matching?

**Comment by rutledj** [comment]
Thks. The source is reading from an Epic hosted drive we are connecting using smb. The destination is writing to a mapped server (located on the same subnet as the mirth appliance occupies). Destination queueing is enabled but it never queues.

Apparently, there is an issue with mapping the Epic drive in the appliance as a network drive. It works for a while and then loses the connection. You have to continually remap the drive.
